doc/pages/expansions.asciidoc

```diff
@@ -1,199 +1,300 @@
+----
+eval set -- $kak_selections
+while [ $# -gt 0 ]; do
+    # ...do a thing with $1...
+    shift
+done
+----
 
-*kak_history_id*::
-    history id of the current buffer, the history id is an integer value
-    which is used to reference a specific buffer version in the undo tree
+The `eval` command will take the expanded `$kak_selections` and unquote them,
+then execute the resulting `set` command, which sets the shell's argument
+variables to the items from `$kak_selections`. The `while` loop with `shift`
+iterates through the arguments one by one.
 
-*kak_runtime*::
-    directory containing the kak support files, determined from kakoune's
-    binary location.
+Only variables actually mentioned in the body of the shell expansion will
+be exported into the shell's environment. For example:
 
-*kak_config*::
-    directory containing the user configuration
+----
+echo %sh{ env | grep ^kak_ }
+----
 
-*kak_version*::
-    version of the current Kakoune server (git hash or release name)
+...will find none of Kakoune's special environment variables, but:
 
-*kak_count*::
-    count parameter passed to the command, defaults to 0 if no count given
+----
+echo %sh{ env | grep ^kak_ # kak_session }
+----
 
-*kak_register*::
-    register parameter passed to the command
+...will find the `$kak_session` variable because it was mentioned by name
+in a comment, even though it wasn't directly used.
 
-*kak_opt_<name>*::
-    value of option *name*
+TIP: These environment variables are also available in other contexts where
+Kakoune uses a shell command, such as the `|`, `!` or `$` normal mode commands
+(See <<keys#,`:doc keys`>>).
 
-*kak_reg_<r>*::
-    quoted list of value of register *r*
+== Value expansions
 
-*kak_main_reg_<r>*::
-    content of register *r* associated with the main selection.
+Expansions with the type `val` give access to Kakoune internal data that is
+not stored in an option or a register. Some value expansions can only be used
+in certain contexts, like `%val{hook_param}` that expands to the parameter
+string of the currently-executing hook, and is not available outside a hook.
 
-*kak_session*::
-    name of the current session
+The following expansions are supported (with required context _in italics_):
 
-*kak_client*::
-    name of the current client
+*%val{buffile}*::
+    _in buffer scope_ +
+    full path of the file or same as `%val{bufname}` when there’s no
+    associated file
 
-*kak_client_pid*::
-    process id of the current client
+*%val{buf_line_count}*::
+    _in buffer scope_ +
+    number of lines in the current buffer
 
-*kak_client_list*::
-    unquoted list of the names of clients connected to the current session
+*%val{buflist}*::
+    quoted list of the names of currently-open buffers (as seen in
+    `%val{bufname}`)
 
-*kak_source*::
-    path of the file currently getting executed (through the source
-    command)
+*%val{bufname}*::
+    _in buffer scope_ +
```

Comment:
There's a paragraph at the end that mentions anything "in buffer scope" is also available in window scope. Maybe it would be better to just say "in buffer, window scope" for such expansions?